internal/util/errors.go

```diff
@@ -16,98 +16,46 @@ limitations under the License.
 
 package util
 
-// ErrKeyNotFound is returned when requested key in omap is not found
-type ErrKeyNotFound struct {
-	keyName string
-	err     error
-}
-
-// NewErrKeyNotFound returns a new ErrKeyNotFound error.
-func NewErrKeyNotFound(keyName string, err error) ErrKeyNotFound {
-	return ErrKeyNotFound{keyName, err}
-}
-
-// Error returns the error string for ErrKeyNotFound.
-func (e ErrKeyNotFound) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrKeyNotFound) Unwrap() error {
-	return e.err
-}
-
-// ErrObjectExists is returned when named omap is already present in rados
-type ErrObjectExists struct {
-	objectName string
-	err        error
-}
-
-// Error returns the error string for ErrObjectExists.
-func (e ErrObjectExists) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrObjectExists) Unwrap() error {
-	return e.err
-}
-
-// ErrObjectNotFound is returned when named omap is not found in rados
-type ErrObjectNotFound struct {
-	oMapName string
-	err      error
-}
-
-// Error returns the error string for ErrObjectNotFound.
-func (e ErrObjectNotFound) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrObjectNotFound) Unwrap() error {
-	return e.err
-}
-
-// ErrSnapNameConflict is generated when a requested CSI snap name already exists on RBD but with
-// different properties, and hence is in conflict with the passed in CSI volume name
-type ErrSnapNameConflict struct {
-	requestName string
-	err         error
-}
+import (
+	"errors"
+	"fmt"
+)
 
-// Error returns the error string for ErrSnapNameConflict.
-func (e ErrSnapNameConflict) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrSnapNameConflict) Unwrap() error {
-	return e.err
-}
+var (
+	// ErrKeyNotFound is returned when requested key in omap is not found
+	ErrKeyNotFound = errors.New("key not found")
+	// ErrObjectExists is returned when named omap is already present in rados
+	ErrObjectExists = errors.New("object exists")
+	// ErrObjectNotFound is returned when named omap is not found in rados
+	ErrObjectNotFound = errors.New("object not found")
+	// ErrSnapNameConflict is generated when a requested CSI snap name already exists on RBD but with
+	// different properties, and hence is in conflict with the passed in CSI volume name
+	ErrSnapNameConflict = errors.New("snapshot name conflict")
+	// ErrPoolNotFound is returned when pool is not found
+	ErrPoolNotFound = errors.New("pool not found")
+)
 
-// NewErrSnapNameConflict returns a ErrSnapNameConflict error when CSI snap name already exists.
-func NewErrSnapNameConflict(name string, err error) ErrSnapNameConflict {
-	return ErrSnapNameConflict{name, err}
+type errPair struct {
+	first  error
```

Comment:
can you add a comment what is the first error and second error? what we get more errors returns from multiple internal functions?

Reply:
Can you rephrase the second sentence? I don't understand it. The `first` and `second` don't have deeper semantics, it's how pair members are named in [C++](http://www.cplusplus.com/reference/utility/pair/), and I just used that. I could also call it `left` and `right` like in binary tree nodes. I have comments later with the exported `Is()` and `Unwrap()` methods, how they treat `first` and `second` differently. But a `errors.Is(myJoinedError, MySentinelError)` would always automatically look into both `first` and `second` errors, if it can find `MySentinelError` there, if this is your question.

Reply:
I would probably go for `inner` and `outer` error, so that `Unwrap()` returns the inner one. Thoughts?

Reply:
For me inner and outer would suggest that there is just a linear wrapping, but in fact it's a bifurcation, so you could unwrap the inner _and_ outer errors. (In my first version I also had a `UnwrapFirst()` function, to make it accessible, but then I changed my mind and wanted to wait until we really need that before adding it.)

Reply:
Yes, both errors are equal in the ErrorPair. Let's put it differently... In many use-cases I expect that one of the errors has a higher priority, and that one should get returned with `Unwrap()`. Users of `JoinErrors()` would normally not call `Unwrap()` themselves, but use `errors.As()` instead. No need for additional functions that will likely not get used.

It would be good to use terminology that gives some understanding which error (the higher priority one) gets returned in `Unwrap()`.

I don't really care how they are called, but `left` and `right` does not give a hint what `Unwrap()` would return.